src/app/blog/multipath-will-fix-this/page.mdx

```diff
@@ -0,0 +1,316 @@
+It forwards UDP datagrams from one endpoint to another.  Since iroh
+endpoints are identified by a [`NodeId`] it means you send it a
+destination `NodeId` together with a datagram.  The relay server might
+now either:
+
+- Drop the datagram on the floor, because the destination endpoint is
+  not connected to this relay server.
+  
+- Forward the datagram to the destination.
+
+The relay server does not need to know what is in the datagram.  In
+fact iroh makes sure it **does not** know what is inside: the payload
+is always encrypted to the destination endpoint [^1].
+
+[^1]: Almost, QUIC's handshake has to establish a TLS connection.
+    This means it has to send the TLS `ClientHello` message in clear
+    text like any other TLS connection on the internet.
+    
+    
+# Holpunching
+
+UDP holepunching is simple really [^simplehp].  All you need is for each
+endpoint to send a UDP datagram to the other at the same time.  The
+NAT routers will think the incoming datagrams are a response to the
+outgoing ones and treat it as a connection.  Now you have a
+holepunched, direct connection.
+
+[^simplehp]: Of course it isn't.  But as already said, the word count
+    of this post is finite.
+
+To do this an endpoint needs to:
+
+- Know which IP addresses it might be reachable on.  Some time we'll
+  write this up in it's own blog post, for now I'll just assume then
+  endpoints know.
+
+- Send these IP address candidates to the remote endpoint via the
+  relay server.
+
+- Once both endpoints have the peer's candidate addresses, send "ping"
+  datagrams to each candidate address of the peer.  Both at the same
+  time.
+  
+- If a "ping" datagram is received, respond with "yay, we
+  holepunched!".  Typically this will be only on 1 IP path out of all
+  the candidates.  Or maybe more and more these days it'll succeed for
+  both an IPv4 and and IPv6 path.
+  
+If you followed arefully you'll have counted 3 special messages that
+need to be sent to the peer endpoint:
+
+1. IP address candidates.  These are sent via the relay server.
+
+2. Pings.  These are sent on the non-relayed IP paths.
+
+3. Pongs.  These are also sent on the non-relayed IP paths.
+
+They need to be sent as UDP datagrams.  Over the same *paths* as the
+QUIC datagrams are also being sent: the **relay path** and any
+**direct paths**.
+
+
+# Multiplexing UDP datagrams
+
+Iroh stands on the shoulders of giants, and it looked carefully at
+ZeroTier and Tailscale.  In particular it borrowed a lot from the DERP
+design from Tailscale.  From the above holepunching description we get
+two kinds of packets:
+
+- Application payload.  For iroh these are QUIC datagrams.
+- Holepunching datagrams.
+
+When an iroh endpoint receives a packet it needs to first figure out
+which kind of packet this is: a QUIC datagram, or a DERP datagram?  If
+it is a QUIC packet it is passed onto the QUIC stack [^quicstack].  If
+it is a DERP datagram it needs to be handed by iroh itself, by a
```

Comment:
Maybe we call them "holepunching datagram" instead of "DERP datagram", to match the list above.